nxontology_ml/model.py

```diff
@@ -0,0 +1,152 @@
+
+    # Filter nodes from the ontology that are labelled
+    for record in islice(_labelled_data(), take):
+        id_ = record.efo_otar_slim_id
+        if id_ in nodes:
+            metrics = nxo.node_info(id_).get_metrics()
+            metrics["__label__"] = record.rs_classification
+            yield metrics
+
+
+@dataclass
+class TrainingData:
+    train_data: FeaturesData
+    train_labels: np.ndarray
+    test_data: FeaturesData
+    test_labels: np.ndarray
+
+
+def to_features_data(
+    records: Iterable[dict[str, Any]],
+    train_frac: float = 0.85,  # 85% of data for training, 15% for testing
+) -> TrainingData:
+    # To pandas
+    df = pd.DataFrame.from_records(records)
+    df = df.sample(frac=1)  # Shuffle
+    training_set_size = int(len(df) * train_frac)
+
+    # Split features by type
+
+    num_feature_names = []
+    num_feature_data = []
+    cat_feature_names = []
+    cat_feature_data = []
+    labels = []
+
+    for feature_name in df:
+        feature_type = FeatureType.from_feature_name(feature_name)
+        if feature_type == FeatureType.LABEL:
+            labels = df[feature_name].to_numpy()
+        elif feature_type == FeatureType.NUMERICAL:
+            num_feature_names.append(feature_name)
+            num_feature_data.append(df[feature_name].to_numpy(dtype=np.float32))
+        elif feature_type == FeatureType.CATEGORICAL:
+            cat_feature_names.append(feature_name)
+            d = [str(x).encode() for x in df[feature_name]]
+            cat_feature_data.append(np.array(d, dtype=object))
+
+    num_feature_data = np.stack(num_feature_data).transpose()
+    cat_feature_data = np.stack(cat_feature_data).transpose()
+
+    return TrainingData(
+        train_data=FeaturesData(
+            num_feature_data=num_feature_data[:training_set_size],
+            cat_feature_data=cat_feature_data[:training_set_size],
+            num_feature_names=num_feature_names,
+            cat_feature_names=cat_feature_names,
+        ),
+        train_labels=labels[:training_set_size],
+        test_data=FeaturesData(
+            num_feature_data=num_feature_data[training_set_size:],
+            cat_feature_data=cat_feature_data[training_set_size:],
+            num_feature_names=num_feature_names,
+            cat_feature_names=cat_feature_names,
+        ),
+        test_labels=labels[training_set_size:],
+    )
+
+
+def train_model(training_data: TrainingData) -> None:
+    train_pool = Pool(data=training_data.train_data, label=training_data.train_labels)
+    test_pool = Pool(data=training_data.test_data, label=training_data.test_labels)
+    model: CatBoostClassifier = CatBoostClassifier(metric_period=100)
+    model.fit(train_pool)
+
+    print("> Feature importance:")
+    print(model.get_feature_importance(test_pool, prettified=True))
+    print()
+    print("> Classification report:")
+    print(classification_report(training_data.test_labels, model.predict(test_pool)))
```

Comment:
Are these predictions on samples that the model has been trained on? Is there any sort of cross-validation or regularization to prevent overfitting?

Reply:
No, by default it's trained on 85% of the data and the evaluation is run on the remaining 15%